lib/rubocop/yaml_duplication_checker.rb

```diff
@@ -5,37 +5,30 @@ module RuboCop
   # @api private
   module YAMLDuplicationChecker
     def self.check(yaml_string, filename, &on_duplicated)
-      # Ruby 2.6+
-      tree = if Gem::Version.new(Psych::VERSION) >= Gem::Version.new('3.1.0')
-               # Specify filename to display helpful message when it raises
-               # an error.
-               YAML.parse(yaml_string, filename: filename)
-             else
-               YAML.parse(yaml_string, filename)
-             end
-      return unless tree
-
-      traverse(tree, &on_duplicated)
-      tree
+      handler = DuplicationCheckHandler.new(&on_duplicated)
+      parser = Psych::Parser.new(handler)
+      parser.parse(yaml_string, filename)
+      parser.handler.root.children[0]
```

Comment:
`Psych::Handlers::DocumentStream` would make this a bit nicer  (it's what `Psych.parse` uses under the hood) but it's marked nodoc so I didn't check it out further.